// @warp-drive/build-config/dist/debugging-BzogyWJo.js
/**
 * ## Debugging
 *
 * Many portions of the internals are helpfully instrumented with logging that can be activated
 * at build time. This instrumentation is always removed from production builds or any builds
 * that has not explicitly activated it. To activate it set the appropriate flag to `true`.
 *
  @module @warp-drive/build-config/debugging
  @main @warp-drive/build-config/debugging
 */
/**
 *
 * Many portions of the internals are helpfully instrumented with logging that can be activated
at build time. This instrumentation is always removed from production builds or any builds
that has not explicitly activated it. To activate it set the appropriate flag to `true`.

```ts
  let app = new EmberApp(defaults, {
    emberData: {
      debug: {
          LOG_PAYLOADS: false, // data store received to update cache with
          LOG_OPERATIONS: false, // updates to cache remote state
          LOG_MUTATIONS: false, // updates to cache local state
          LOG_NOTIFICATIONS: false,
          LOG_REQUESTS: false,
          LOG_REQUEST_STATUS: false,
          LOG_IDENTIFIERS: false,
          LOG_GRAPH: false,
          LOG_INSTANCE_CACHE: false,
      }
    }
  });
  ```

  @class DebugLogging
  @public
 */
/**
 * log payloads received by the store
 * via `push` or returned from a delete/update/create
 * operation.
 *
 * @property {boolean} LOG_PAYLOADS
 * @public
 */
const LOG_PAYLOADS = false;
/**
 * log remote-state updates to the cache
 *
 * @property {boolean} LOG_OPERATIONS
 * @public
 */
const LOG_OPERATIONS = false;
/**
 * log local-state updates to the cache
 *
 * @property {boolean} LOG_MUTATIONS
 * @public
 */
const LOG_MUTATIONS = false;
/**
 * log notifications received by the NotificationManager
 *
 * @property {boolean} LOG_NOTIFICATIONS
 * @public
 */
const LOG_NOTIFICATIONS = false;
/**
 * log requests issued by the RequestManager
 *
 * @property {boolean} LOG_REQUESTS
 * @public
 */
const LOG_REQUESTS = false;
/**
 * log updates to requests the store has issued to
 * the network (adapter) to fulfill.
 *
 * @property {boolean} LOG_REQUEST_STATUS
 * @public
 */
const LOG_REQUEST_STATUS = false;
/**
 * log peek, generation and updates to
 * Record Identifiers.
 *
 * @property {boolean} LOG_IDENTIFIERS
 * @public
 */
const LOG_IDENTIFIERS = false;
/**
 * log updates received by the graph (relationship pointer storage)
 *
 * @property {boolean} LOG_GRAPH
 * @public
 */
const LOG_GRAPH = false;
/**
 * log creation/removal of RecordData and Record
 * instances.
 *
 * @property {boolean} LOG_INSTANCE_CACHE
 * @public
 */
const LOG_INSTANCE_CACHE = false;

const LOGGING = /*#__PURE__*/Object.freeze(/*#__PURE__*/Object.defineProperty({
  __proto__: null,
  LOG_GRAPH,
  LOG_IDENTIFIERS,
  LOG_INSTANCE_CACHE,
  LOG_MUTATIONS,
  LOG_NOTIFICATIONS,
  LOG_OPERATIONS,
  LOG_PAYLOADS,
  LOG_REQUESTS,
  LOG_REQUEST_STATUS
}, Symbol.toStringTag, { value: 'Module' }));

export { LOGGING as L, LOG_PAYLOADS as a, LOG_OPERATIONS as b, LOG_MUTATIONS as c, LOG_NOTIFICATIONS as d, LOG_REQUESTS as e, LOG_REQUEST_STATUS as f, LOG_IDENTIFIERS as g, LOG_GRAPH as h, LOG_INSTANCE_CACHE as i };
