// @warp-drive/build-config/dist/deprecations-ChFQtx-4.js
// deprecations
const DEPRECATE_CATCH_ALL = true;
const DEPRECATE_COMPUTED_CHAINS = true;
const DEPRECATE_NON_STRICT_TYPES = true;
const DEPRECATE_NON_STRICT_ID = true;
const DEPRECATE_LEGACY_IMPORTS = true;
const DEPRECATE_NON_UNIQUE_PAYLOADS = true;
const DEPRECATE_RELATIONSHIP_REMOTE_UPDATE_CLEARING_LOCAL_STATE = true;
const DEPRECATE_MANY_ARRAY_DUPLICATES = true;
const DEPRECATE_STORE_EXTENDS_EMBER_OBJECT = true;
const ENABLE_LEGACY_SCHEMA_SERVICE = true;
const DEPRECATE_EMBER_INFLECTOR = true;
const DISABLE_7X_DEPRECATIONS = true;

const CURRENT_DEPRECATIONS = /*#__PURE__*/Object.freeze(/*#__PURE__*/Object.defineProperty({
	__proto__: null,
	DEPRECATE_CATCH_ALL,
	DEPRECATE_COMPUTED_CHAINS,
	DEPRECATE_EMBER_INFLECTOR,
	DEPRECATE_LEGACY_IMPORTS,
	DEPRECATE_MANY_ARRAY_DUPLICATES,
	DEPRECATE_NON_STRICT_ID,
	DEPRECATE_NON_STRICT_TYPES,
	DEPRECATE_NON_UNIQUE_PAYLOADS,
	DEPRECATE_RELATIONSHIP_REMOTE_UPDATE_CLEARING_LOCAL_STATE,
	DEPRECATE_STORE_EXTENDS_EMBER_OBJECT,
	DISABLE_7X_DEPRECATIONS,
	ENABLE_LEGACY_SCHEMA_SERVICE
}, Symbol.toStringTag, { value: 'Module' }));

export { CURRENT_DEPRECATIONS as C, DEPRECATE_CATCH_ALL as D, ENABLE_LEGACY_SCHEMA_SERVICE as E, DEPRECATE_COMPUTED_CHAINS as a, DEPRECATE_NON_STRICT_TYPES as b, DEPRECATE_NON_STRICT_ID as c, DEPRECATE_LEGACY_IMPORTS as d, DEPRECATE_NON_UNIQUE_PAYLOADS as e, DEPRECATE_RELATIONSHIP_REMOTE_UPDATE_CLEARING_LOCAL_STATE as f, DEPRECATE_MANY_ARRAY_DUPLICATES as g, DEPRECATE_STORE_EXTENDS_EMBER_OBJECT as h, DEPRECATE_EMBER_INFLECTOR as i, DISABLE_7X_DEPRECATIONS as j };
