// ember-file-upload/dist/data-transfer-wrapper-59b09dd4.js
import { a as _defineProperty } from './rate-6e913d98.js';

const getDataSupport = {};
class DataTransferWrapper {
  constructor(event) {
    _defineProperty(this, "dataTransfer", void 0);
    _defineProperty(this, "itemDetails", void 0);
    _defineProperty(this, "source", void 0);
    this.source = event.source;
    this.dataTransfer = event.dataTransfer;
    this.itemDetails = event.itemDetails;
  }
  getData(type) {
    const dataTransfer = this.dataTransfer;
    if (!dataTransfer) return;

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    if (getDataSupport[type] == null) {
      try {
        const data = dataTransfer.getData(type);
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        getDataSupport[type] = true;
        return data;
      } catch (e) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        getDataSupport[type] = false;
      }
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
    } else if (getDataSupport[type]) {
      return dataTransfer.getData(type);
    }
    return '';
  }
  get filesOrItems() {
    return this.files.length ? this.files : this.items;
  }
  get files() {
    return Array.from(this.dataTransfer?.files ?? []);
  }
  get items() {
    return this.itemDetails ?? Array.from(this.dataTransfer?.items ?? []);
  }
}

export { DataTransferWrapper as D };
